File: MIDTERM-LAB-3/MP07/MP07_SortRecords.java
// ----------------------------------------------------------
// MP07  |  Sort Records Alphabetically
// Programmer  : NAGPATON, FRANCESCA LOUISE MAY G.
// Course      : Programming 2 — BSIT GD 1st Year
// Description : Reads a CSV file of Pearson VUE exam results
//               and displays all records sorted A–Z by name.
// ----------------------------------------------------------
import java.io.*;
import java.util.*;

public class MP07_SortRecords {

    // Represents one row/record from the CSV file
    static class Record {
        String lastName;   // Candidate's last names
        String firstName;  // Candidate's first name
        String type;       // Student / Faculty / NTE
        String exam;       // Exam name
        String language;   // Exam language
        String date;       // Exam date
        int score;         // Exam score
        String result;     // PASS or FAIL
        String timeUsed;   // Time taken

        // Constructor to initialize all fields of a record
        Record(String lastName, String firstName, String type,
               String exam, String language, String date,
               int score, String result, String timeUsed) {
            this.lastName  = lastName;
            this.firstName = firstName;
            this.type      = type;
            this.exam      = exam;
            this.language  = language;
            this.date      = date;
            this.score     = score;
            this.result    = result;
            this.timeUsed  = timeUsed;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Prompt user for the CSV file path
        System.out.print("Enter CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        List<Record> records = new ArrayList<>(); // List to store all parsed records

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            boolean headerFound = false; // Flag to detect the header row

            while ((line = br.readLine()) != null) {
                // Skip blank lines
                if (line.trim().isEmpty()) continue;

                // Detect and skip the header row
                if (line.contains("Candidate") && line.contains("Exam") && line.contains("Score")) {
                    headerFound = true;
                    continue;
                }

                // Skip metadata/title rows before the header
                if (!headerFound) continue;

                // Parse each CSV line using a proper comma-aware split
                String[] cols = parseCSVLine(line);
                if (cols.length < 8) continue; // Skip malformed rows

                // The name field is "LastName,FirstName" (stored in col 0)
                String[] nameParts = cols[0].split(",", 2);
                if (nameParts.length < 2) continue;

                String lastName  = nameParts[0].trim();
                String firstName = nameParts[1].trim();
                String type      = cols[1].trim();   // Student/Faculty/NTE
                // cols[2] is empty (Column1 placeholder)
                String exam      = cols[3].trim();   // Exam name
                String language  = cols[4].trim();   // Language
                String date      = cols[5].trim();   // Date
                int score;
                try {
                    score = Integer.parseInt(cols[6].trim()); // Score
                } catch (NumberFormatException e) {
                    continue; // Skip rows with invalid score
                }
                String result   = cols[7].trim();    // PASS or FAIL
                String timeUsed = cols.length > 8 ? cols[8].trim() : ""; // Time used

                records.add(new Record(lastName, firstName, type,
                        exam, language, date, score, result, timeUsed));
            }
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }

        // Sort records alphabetically by last name, then first name
        records.sort(Comparator.comparing((Record r) -> r.lastName.toLowerCase())
                               .thenComparing(r -> r.firstName.toLowerCase()));

        // Display header
        System.out.println("\n============================================================");
        System.out.println("           MP07 - RECORDS SORTED ALPHABETICALLY             ");
        System.out.println("============================================================");
        System.out.printf("%-5s %-20s %-20s %-10s %-5s %-6s%n",
                "No.", "Last Name", "First Name", "Type", "Score", "Result");
        System.out.println("------------------------------------------------------------");

        // Display each sorted record
        int count = 1;
        for (Record r : records) {
            System.out.printf("%-5d %-20s %-20s %-10s %-5d %-6s%n",
                    count++, r.lastName, r.firstName, r.type, r.score, r.result);
        }

        System.out.println("------------------------------------------------------------");
        System.out.println("Total Records: " + records.size());
    }

    /**
     * Parses a single CSV line, correctly handling quoted fields
     * that may contain commas (e.g., "LastName,FirstName").
     * @param line - raw CSV line string
     * @return array of field values
     */
    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        boolean inQuotes = false;   // Tracks if currently inside a quoted field
        StringBuilder sb = new StringBuilder();

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes; // Toggle quote mode
            } else if (c == ',' && !inQuotes) {
                fields.add(sb.toString()); // End of a field
                sb.setLength(0);           // Reset builder
            } else {
                sb.append(c);
            }
        }
        fields.add(sb.toString()); // Add last field
        return fields.toArray(new String[0]);
    }
}

File: MIDTERM-LAB-3/MP07/MP07_SortRecords.js
/**
 * MP07 - Sort Records Alphabetically
 * Reads a CSV dataset and sorts candidate records
 * alphabetically by last name, then displays the sorted results.
 */

const fs = require("fs");
const readline = require("readline");

// Interface to read user input from the terminal
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Prompt the user to enter the CSV file path
rl.question("Enter CSV dataset file path: ", (filePath) => {
  rl.close();

  // Read the file content
  let content;
  try {
    content = fs.readFileSync(filePath.trim(), "utf8");
  } catch (err) {
    console.log("Error reading file: " + err.message);
    return;
  }

  const lines = content.split(/\r?\n/); // Split into individual lines
  const records = [];                   // Array to store parsed records
  let headerFound = false;              // Flag to locate the data header row

  for (const line of lines) {
    // Skip blank lines
    if (!line.trim()) continue;

    // Detect the header row
    if (line.includes("Candidate") && line.includes("Exam") && line.includes("Score")) {
      headerFound = true;
      continue;
    }

    // Skip metadata rows before header
    if (!headerFound) continue;

    // Parse the CSV line (handles quoted commas inside names)
    const cols = parseCSVLine(line);
    if (cols.length < 8) continue; // Skip malformed rows

    // Name field is formatted as "LastName,FirstName"
    const nameParts = cols[0].split(",");
    if (nameParts.length < 2) continue;

    const lastName  = nameParts[0].trim();
    const firstName = nameParts[1].trim();
    const type      = cols[1].trim();   // Student/Faculty/NTE
    const exam      = cols[3].trim();   // Exam name
    const language  = cols[4].trim();   // Language
    const date      = cols[5].trim();   // Date
    const score     = parseInt(cols[6].trim()); // Score (numeric)
    const result    = cols[7].trim();   // PASS or FAIL
    const timeUsed  = cols[8] ? cols[8].trim() : ""; // Time used

    if (isNaN(score)) continue; // Skip rows with invalid score

    records.push({ lastName, firstName, type, exam, language, date, score, result, timeUsed });
  }

  // Sort alphabetically by last name, then first name (case-insensitive)
  records.sort((a, b) => {
    const lastCmp = a.lastName.toLowerCase().localeCompare(b.lastName.toLowerCase());
    if (lastCmp !== 0) return lastCmp;
    return a.firstName.toLowerCase().localeCompare(b.firstName.toLowerCase());
  });

  // Display the sorted results
  console.log("\n============================================================");
  console.log("           MP07 - RECORDS SORTED ALPHABETICALLY             ");
  console.log("============================================================");
  console.log(
    padRight("No.", 5) +
    padRight("Last Name", 20) +
    padRight("First Name", 20) +
    padRight("Type", 10) +
    padRight("Score", 6) +
    "Result"
  );
  console.log("------------------------------------------------------------");

  records.forEach((r, i) => {
    console.log(
      padRight(String(i + 1), 5) +
      padRight(r.lastName, 20) +
      padRight(r.firstName, 20) +
      padRight(r.type, 10) +
      padRight(String(r.score), 6) +
      r.result
    );
  });

  console.log("------------------------------------------------------------");
  console.log("Total Records: " + records.length);
});

/**
 * Parses a single CSV line, correctly handling quoted fields
 * that may contain commas (e.g., "LastName,FirstName").
 * @param {string} line - raw CSV line
 * @returns {string[]} array of field values
 */
function parseCSVLine(line) {
  const fields = [];
  let inQuotes = false; // Tracks if inside a quoted field
  let current  = "";

  for (const ch of line) {
    if (ch === '"') {
      inQuotes = !inQuotes; // Toggle quoted mode
    } else if (ch === "," && !inQuotes) {
      fields.push(current); // Save current field
      current = "";          // Reset for next field
    } else {
      current += ch;
    }
  }
  fields.push(current); // Push last field
  return fields;
}

/**
 * Pads a string to a fixed width for aligned console output.
 * @param {string} str - the string to pad
 * @param {number} width - desired width
 * @returns {string} padded string
 */
function padRight(str, width) {
  return str.padEnd(width);
}